contracts/base-staking/CompoundRateKeeperV2.sol

```diff
@@ -0,0 +1,186 @@
+        capitalizationPeriod = _capitalizationPeriod;
+
+        lastUpdate = uint64(block.timestamp);
+
+        emit CapitalizationPeriodChanged(_capitalizationPeriod);
+    }
+
+    /// @dev Call this function only when getCompoundRate() or getPotentialCompoundRate()
+    function emergencyUpdateCompoundRate() external override {
+        try this.getCompoundRate() returns (uint256 _rate) {
+            if (_rate == __getMaxRate()) hasMaxRateReached = true;
+        } catch {
+            hasMaxRateReached = true;
+        }
+    }
+
+    /// @dev @notice Calculate compound rate for this moment.
+    function getCompoundRate() public view override returns (uint256) {
+        return __getPotentialCompoundRate(uint64(block.timestamp));
+    }
+
+    /**
+     * @dev Calculate compound rate at a particular time.
+     *
+     * Main contract logic, calculate actual compound rate.
+     * If rate bigger than __getMaxRate(), return __getMaxRate().
+     * If function is reverted by overflow, call emergencyUpdateCompoundRate()
+     */
+    function getPotentialCompoundRate(uint64 _timestamp) public view override returns (uint256) {
+        return __getPotentialCompoundRate(_timestamp);
+    }
+
+    function _getDecimals() internal pure returns (uint256) {
+        return 10**27;
+    }
+
+    /// @dev Max accessible compound rate.
+    function __getMaxRate() private pure returns (uint256) {
+        return type(uint128).max * _getDecimals();
+    }
+
+    function __getPotentialCompoundRate(uint64 _timestamp) private view returns (uint256) {
+        if (hasMaxRateReached) return __getMaxRate();
+
+        uint64 _lastUpdate = lastUpdate;
+
+        // Require is made to avoid incorrect calculations at the front
+        require(_lastUpdate <= _timestamp, "CRK: invalid timestamp");
+
+        if (_timestamp == _lastUpdate) return currentRate;
+
+        uint64 _secondsPassed = _timestamp - _lastUpdate;
+
+        uint64 _capitalizationPeriod = capitalizationPeriod;
+        uint64 _capitalizationPeriodsNum = _secondsPassed / _capitalizationPeriod;
+        uint64 _secondsLeft = _secondsPassed % _capitalizationPeriod;
+
+        uint256 _annualPercent = annualPercent;
+        uint256 _rate = currentRate;
+
+        if (_capitalizationPeriodsNum != 0) {
+            uint256 _capitalizationPeriodRate = __pow(
+                _annualPercent,
+                _capitalizationPeriodsNum,
+                _getDecimals()
+            );
+            _rate = (_rate * _capitalizationPeriodRate) / _getDecimals();
+        }
+
+        if (_secondsLeft > 0) {
+            uint256 _rateLeft = _getDecimals() +
+                ((_annualPercent - _getDecimals()) * _secondsLeft) /
+                _capitalizationPeriod;
+            _rate = (_rate * _rateLeft) / _getDecimals();
+        }
+
+        return _rate.min(__getMaxRate());
+    }
+
+    /// @dev github.com/makerdao/dss implementation of exponentiation by squaring.
+    function __pow(
```

Comment:
This function will be added to the modules themselves. Waiting for the fixes from @tynukua